File: fix_flv_stream.py
#!/usr/bin/env python3
import argparse
import mmap
import re
import struct
import subprocess
import sys
import tempfile
import hashlib
from collections import Counter
from dataclasses import dataclass
from pathlib import Path


@dataclass
class Tag:
    type: int
    size: int
    timestamp: int
    stream_id: int
    offset: int
    payload: bytes


def parse_aac_sample_rate(payload):
    if len(payload) < 4:
        return None
    frequencies = [
        96000, 88200, 64000, 48000, 44100, 32000, 24000,
        22050, 16000, 12000, 11025, 8000, 7350,
    ]
    frequency_index = ((payload[2] & 0x07) << 1) | (payload[3] >> 7)
    if frequency_index >= len(frequencies):
        return None
    return frequencies[frequency_index]


def parse_composition_offset(payload):
    value = int.from_bytes(payload[2:5], "big")
    if value & 0x800000:
        value -= 0x1000000
    return value


def update_progress(label, position, total, last_percent):
    percent = 0 if total == 0 else min(100, position * 100 // total)
    if percent != last_percent:
        print(f"\r{label}: {percent:3d}%", end="", file=sys.stderr, flush=True)
    return percent


def finish_progress(label):
    print(f"\r{label}: 100%", file=sys.stderr, flush=True)


def iter_tags(path, progress_label=None):
    last_percent = -1
    with path.open("rb") as source_file:
        with mmap.mmap(source_file.fileno(), 0, access=mmap.ACCESS_READ) as mapped:
            size_limit = len(mapped) - 15
            for match in re.finditer(rb"[\x08\x09\x12]", mapped):
                offset = match.start()
                if progress_label:
                    last_percent = update_progress(
                        progress_label, offset, len(mapped), last_percent
                    )
                if offset > size_limit:
                    continue
                tag_type = mapped[offset]
                tag_size = int.from_bytes(mapped[offset + 1:offset + 4], "big")
                if tag_size <= 0 or tag_size > 10_000_000 or offset + 15 + tag_size > len(mapped):
                    continue
                previous_size = int.from_bytes(
                    mapped[offset + 11 + tag_size:offset + 15 + tag_size],
                    "big",
                )
                if previous_size != tag_size + 11:
                    continue
                low_timestamp = int.from_bytes(mapped[offset + 4:offset + 7], "big")
                high_timestamp = int.from_bytes(mapped[offset + 7:offset + 8], "big")
                timestamp = (high_timestamp << 24) | low_timestamp
                stream_id = int.from_bytes(mapped[offset + 8:offset + 11], "big")
                payload = mapped[offset + 11:offset + 11 + tag_size]
                yield Tag(tag_type, tag_size, timestamp, stream_id, offset, payload)
            if progress_label:
                finish_progress(progress_label)


def inspect_tags(path):
    audio_packets = 0
    video_packets = 0
    sample_rate = None
    has_audio_config = False
    has_video_config = False
    audio_keys = []
    video_keys = []

    for tag in iter_tags(path, "scanning"):
        payload = tag.payload
        if tag.type == 8 and payload and payload[0] >> 4 == 10:
            if len(payload) > 1 and payload[1] == 0:
                has_audio_config = True
                sample_rate = sample_rate or parse_aac_sample_rate(payload)
            elif len(payload) > 1 and sample_rate:
                audio_packets += 1
                audio_keys.append((hashlib.sha1(payload).digest(), tag.timestamp))
        elif tag.type == 9 and payload and payload[0] & 0x0F == 7:
            if len(payload) > 1 and payload[1] == 0:
                has_video_config = True
            elif len(payload) >= 5:
                video_packets += 1
                video_keys.append((hashlib.sha1(payload).digest(), tag.timestamp))

    audio_prefix_skip = duplicate_prefix_count(audio_keys)
    video_prefix_skip = duplicate_prefix_count(video_keys)
    audio_packets -= audio_prefix_skip
    video_packets -= video_prefix_skip
    base = min(
        audio_keys[audio_prefix_skip][1],
        video_keys[video_prefix_skip][1],
    )

    return {
        "audio_packets": audio_packets,
        "video_packets": video_packets,
        "sample_rate": sample_rate,
        "has_audio_config": has_audio_config,
        "has_video_config": has_video_config,
        "audio_prefix_skip": audio_prefix_skip,
        "video_prefix_skip": video_prefix_skip,
        "base": base,
    }


def duplicate_prefix_count(keys):
    if not keys:
        return 0
    counts = Counter(keys)
    count = 0
    for key in keys:
        if counts[key] < 2:
            break
        count += 1
    return count


def write_tag(output_file, tag_type, timestamp, payload):
    size = len(payload)
    timestamp_bytes = int(timestamp).to_bytes(4, "big")
    output_file.write(
        bytes([tag_type])
        + size.to_bytes(3, "big")
        + timestamp_bytes[1:]
        + timestamp_bytes[0:1]
        + b"\x00\x00\x00"
        + payload
    )
    output_file.write((size + 11).to_bytes(4, "big"))


def write_clean_flv(source, output, info):
    audio_packets = info["audio_packets"]
    video_packets = info["video_packets"]
    sample_rate = info["sample_rate"]
    if not audio_packets or not video_packets or not sample_rate:
        raise RuntimeError("not enough valid audio/video packets")

    wrote_audio_config = False
    wrote_video_config = False
    started = False
    audio_seen = 0
    video_seen = 0
    audio_skip = info["audio_prefix_skip"]
    video_skip = info["video_prefix_skip"]

    with source.open("rb") as source_file, output.open("wb") as output_file:
        output_file.write(b"FLV\x01\x05\x00\x00\x00\t")
        output_file.write(b"\x00\x00\x00\x00")

        for tag in iter_tags(source, "rewriting"):
            payload = tag.payload
            if tag.type == 9 and payload and payload[0] & 0x0F == 7:
                if len(payload) > 1 and payload[1] == 0:
                    if not wrote_video_config:
                        write_tag(output_file, 9, 0, payload)
                        wrote_video_config = True
                elif len(payload) >= 5:
                    if video_seen < video_skip:
                        video_seen += 1
                        continue
                    if not started and payload[0] >> 4 != 1:
                        video_seen += 1
                        continue
                    started = True
                    timestamp = max(0, tag.timestamp - info["base"])
                    write_tag(output_file, 9, timestamp, payload)
                    video_seen += 1
            elif tag.type == 8 and payload and payload[0] >> 4 == 10:
                if len(payload) > 1 and payload[1] == 0:
                    if not wrote_audio_config:
                        write_tag(output_file, 8, 0, payload)
                        wrote_audio_config = True
                elif len(payload) > 1:
                    if audio_seen < audio_skip:
                        audio_seen += 1
                        continue
                    if not started:
                        audio_seen += 1
                        continue
                    timestamp = max(0, tag.timestamp - info["base"])
                    write_tag(output_file, 8, timestamp, payload)
                    audio_seen += 1


def probe_output(path):
    result = subprocess.run(
        [
            "ffprobe", "-v", "error", "-show_streams", "-of", "json", str(path)
        ],
        text=True,
        stdout=subprocess.PIPE,
        stderr=subprocess.PIPE,
    )
    if result.returncode != 0:
        return None
    import json
    try:
        streams = json.loads(result.stdout).get("streams", [])
        video = next((stream for stream in streams if stream.get("codec_type") == "video"), None)
        audio = next((stream for stream in streams if stream.get("codec_type") == "audio"), None)
        if video is None or audio is None:
            return None
        return float(video.get("duration", 0)), float(audio.get("duration", 0))
    except (json.JSONDecodeError, TypeError, ValueError):
        return None


def repair(source, output):
    info = inspect_tags(source)
    print(
        f"valid audio packets: {info['audio_packets']}, "
        f"valid video packets: {info['video_packets']}, "
        f"sample rate: {info['sample_rate']}"
    )

    with tempfile.TemporaryDirectory(dir=output.parent) as temp_dir:
        clean_flv = Path(temp_dir) / f"{source.stem}.clean.flv"
        write_clean_flv(source, clean_flv, info)

        command = [
            "ffmpeg",
            "-hide_banner",
            "-v",
            "warning",
            "-stats",
            "-i",
            str(clean_flv),
            "-map",
            "0:v:0",
            "-map",
            "0:a:0",
            "-c",
            "copy",
            "-movflags",
            "+faststart",
            "-y",
            str(output),
        ]
        result = subprocess.run(command)
        if result.returncode != 0:
            return False

        durations = probe_output(output)
        if durations is None:
            return False
        video_duration, audio_duration = durations
        print(f"video duration: {video_duration:.3f}s, audio duration: {audio_duration:.3f}s")
        return abs(video_duration - audio_duration) < 0.5


def iter_flv_files(directory):
    return sorted(
        path
        for path in directory.rglob("*")
        if path.is_file() and path.suffix.lower() == ".flv"
    )


def repair_one(source, output, overwrite):
    if output.exists() and not overwrite:
        print(f"skip: {output.name} already exists")
        return True

    try:
        if not repair(source, output):
            if output.exists():
                output.unlink()
            return False
    except Exception as error:
        if output.exists():
            output.unlink()
        print(f"error: {error}", file=sys.stderr)
        return False
    return True


def repair_directory(source, output_root, overwrite):
    output_root.mkdir(parents=True, exist_ok=True)
    files = iter_flv_files(source)
    if not files:
        print("No FLV files found.")
        return 0

    failed = []
    for source_file in files:
        output_file = output_root / f"{source_file.stem}_streamcopy.mp4"
        print(f"\n{source_file}")
        if not repair_one(source_file, output_file, overwrite):
            failed.append(source_file)

    if failed:
        print(f"\n{len(failed)} file(s) failed:")
        for source_file in failed:
            print(f"  {source_file}")
        return 1
    return 0


def main():
    parser = argparse.ArgumentParser(
        description="Stream-parse an FLV, drop invalid tags, and stream-copy A/V into MP4.\n用于修复brec命令报错的flv文件"
    )
    parser.add_argument("source", type=Path)
    parser.add_argument("output", nargs="?", type=Path)
    parser.add_argument("--overwrite", action="store_true")
    args = parser.parse_args()

    source = args.source.expanduser().resolve()
    if source.is_file():
        output = args.output.expanduser().resolve() if args.output else source.with_name(
            f"{source.stem}_streamcopy.mp4"
        )
        return 0 if repair_one(source, output, args.overwrite) else 1

    if source.is_dir():
        output_root = args.output.expanduser().resolve() if args.output else source
        return repair_directory(source, output_root, args.overwrite)

    parser.error(f"not a file or directory: {source}")


if __name__ == "__main__":
    raise SystemExit(main())

File: fix_flv_stream.js
#!/usr/bin/env node
"use strict";

const crypto = require("crypto");
const fs = require("fs");
const fsp = require("fs").promises;
const os = require("os");
const path = require("path");
const { spawn } = require("child_process");

const TAG_TYPE_BYTES = [0x08, 0x09, 0x12];
const CHUNK_SIZE = 4 * 1024 * 1024;

function parseAacSampleRate(payload) {
  if (payload.length < 4) {
    return null;
  }

  const frequencies = [
    96000, 88200, 64000, 48000, 44100, 32000, 24000,
    22050, 16000, 12000, 11025, 8000, 7350,
  ];
  const frequencyIndex = ((payload[2] & 0x07) << 1) | (payload[3] >> 7);
  return frequencyIndex < frequencies.length ? frequencies[frequencyIndex] : null;
}

function updateProgress(label, position, total, lastPercent) {
  const percent = total === 0 ? 0 : Math.min(100, Math.floor((position * 100) / total));
  if (percent !== lastPercent) {
    process.stderr.write(`\r${label}: ${String(percent).padStart(3, " ")}%`);
  }
  return percent;
}

function finishProgress(label) {
  process.stderr.write(`\r${label}: 100%\n`);
}

function findCandidate(buffer, from) {
  let searchFrom = from;
  while (searchFrom < buffer.length) {
    let bestIndex = -1;
    for (const tagTypeByte of TAG_TYPE_BYTES) {
      const index = buffer.indexOf(tagTypeByte, searchFrom);
      if (index !== -1 && (bestIndex === -1 || index < bestIndex)) {
        bestIndex = index;
      }
    }
    if (bestIndex === -1) {
      break;
    }
    return bestIndex;
  }
  return -1;
}

async function iterTags(sourcePath, progressLabel, onTag) {
  const fileHandle = fs.openSync(sourcePath, "r");
  let lastPercent = -1;

  try {
    const stat = fs.fstatSync(fileHandle);
    const total = stat.size;
    const sizeLimit = total - 15;
    const readBuffer = Buffer.allocUnsafe(Math.min(CHUNK_SIZE, total));
    let carry = Buffer.alloc(0);
    let nextScan = 0;

    for (let position = 0; position < total; position += readBuffer.length) {
      const length = Math.min(readBuffer.length, total - position);
      const chunk = readBuffer.subarray(0, length);
      if (fs.readSync(fileHandle, chunk, 0, length, position) !== length) {
        throw new Error(`unexpected end of file: ${sourcePath}`);
      }

      const currentEnd = position + length;
      const buffer = carry.length === 0 ? chunk : Buffer.concat([carry, chunk]);
      const bufferBase = currentEnd - buffer.length;
      const scanStart = Math.max(0, nextScan - bufferBase);

      for (
        let index = findCandidate(buffer, scanStart);
        index !== -1;
        index = findCandidate(buffer, index + 1)
      ) {
        const tagOffset = bufferBase + index;
        if (tagOffset > sizeLimit) {
          continue;
        }

        if (progressLabel) {
          lastPercent = updateProgress(progressLabel, tagOffset, total, lastPercent);
        }

        const tagSize = buffer.readUIntBE(index + 1, 3);
        if (
          tagSize <= 0 ||
          tagSize > 10_000_000 ||
          tagOffset + 15 + tagSize > total
        ) {
          continue;
        }

        const previousIndex = index + 11 + tagSize;
        let previousSize;
        let payload;

        if (previousIndex + 4 <= buffer.length) {
          previousSize = buffer.readUInt32BE(previousIndex);
          payload = buffer.subarray(index + 11, index + 11 + tagSize);
        } else {
          const previousBuffer = Buffer.alloc(4);
          fs.readSync(
            fileHandle,
            previousBuffer,
            0,
            4,
            tagOffset + 11 + tagSize,
          );
          previousSize = previousBuffer.readUInt32BE(0);
          if (previousSize !== tagSize + 11) {
            continue;
          }

          payload = Buffer.allocUnsafe(tagSize);
          if (fs.readSync(fileHandle, payload, 0, tagSize, tagOffset + 11) !== tagSize) {
            continue;
          }
        }

        if (previousSize !== tagSize + 11) {
          continue;
        }

        const lowTimestamp = buffer.readUIntBE(index + 4, 3);
        const highTimestamp = buffer[index + 7];
        const timestamp = (highTimestamp << 24) | lowTimestamp;
        const streamId = buffer.readUIntBE(index + 8, 3);
        await onTag({
          type: buffer[index],
          size: tagSize,
          timestamp: timestamp >>> 0,
          streamId,
          offset: tagOffset,
          payload,
        });
      }

      nextScan = Math.max(nextScan, currentEnd - 15);
      carry = buffer.subarray(Math.max(0, buffer.length - 15));
    }

    if (progressLabel) {
      finishProgress(progressLabel);
    }
  } finally {
    fs.closeSync(fileHandle);
  }
}

function duplicatePrefixCount(keys) {
  if (keys.length === 0) {
    return 0;
  }

  const counts = new Map();
  for (const key of keys) {
    counts.set(key.key, (counts.get(key.key) || 0) + 1);
  }

  let count = 0;
  for (const key of keys) {
    if ((counts.get(key.key) || 0) < 2) {
      break;
    }
    count += 1;
  }
  return count;
}

async function inspectTags(sourcePath) {
  let audioPackets = 0;
  let videoPackets = 0;
  let sampleRate = null;
  let hasAudioConfig = false;
  let hasVideoConfig = false;
  const audioKeys = [];
  const videoKeys = [];

  await iterTags(sourcePath, "scanning", (tag) => {
    const payload = tag.payload;
    if (tag.type === 8 && payload.length > 0 && (payload[0] >> 4) === 10) {
      if (payload.length > 1 && payload[1] === 0) {
        hasAudioConfig = true;
        sampleRate = sampleRate || parseAacSampleRate(payload);
      } else if (payload.length > 1 && sampleRate) {
        audioPackets += 1;
        const digest = crypto.createHash("sha1").update(payload).digest("hex");
        audioKeys.push({ key: `${digest}:${tag.timestamp}`, timestamp: tag.timestamp });
      }
    } else if (tag.type === 9 && payload.length > 0 && (payload[0] & 0x0f) === 7) {
      if (payload.length > 1 && payload[1] === 0) {
        hasVideoConfig = true;
      } else if (payload.length >= 5) {
        videoPackets += 1;
        const digest = crypto.createHash("sha1").update(payload).digest("hex");
        videoKeys.push({ key: `${digest}:${tag.timestamp}`, timestamp: tag.timestamp });
      }
    }
  });

  const audioPrefixSkip = duplicatePrefixCount(audioKeys);
  const videoPrefixSkip = duplicatePrefixCount(videoKeys);
  audioPackets -= audioPrefixSkip;
  videoPackets -= videoPrefixSkip;

  if (audioKeys.length === 0 || videoKeys.length === 0) {
    throw new Error("not enough valid audio/video packets");
  }

  const base = Math.min(
    audioKeys[audioPrefixSkip].timestamp,
    videoKeys[videoPrefixSkip].timestamp,
  );

  return {
    audioPackets,
    videoPackets,
    sampleRate,
    hasAudioConfig,
    hasVideoConfig,
    audioPrefixSkip,
    videoPrefixSkip,
    base,
  };
}

function writeTag(fileHandle, tagType, timestamp, payload) {
  const size = payload.length;
  const timestampBytes = Buffer.alloc(4);
  timestampBytes.writeUInt32BE(timestamp, 0);

  const tag = Buffer.allocUnsafe(11 + size + 4);
  tag[0] = tagType;
  tag.writeUIntBE(size, 1, 3);
  tag.set(timestampBytes.subarray(1, 4), 4);
  tag.set(timestampBytes.subarray(0, 1), 7);
  tag.set(Buffer.from([0, 0, 0]), 8);
  payload.copy(tag, 11);
  tag.writeUInt32BE(size + 11, 11 + size);
  return fileHandle.write(tag);
}

async function writeCleanFlv(sourcePath, outputPath, info) {
  const outputHandle = await fsp.open(outputPath, "w");
  let wroteAudioConfig = false;
  let wroteVideoConfig = false;
  let started = false;
  let audioSeen = 0;
  let videoSeen = 0;

  try {
    await outputHandle.write(Buffer.from("FLV\x01\x05\x00\x00\x00\t", "binary"));
    await outputHandle.write(Buffer.from([0, 0, 0, 0]));

    await iterTags(sourcePath, "rewriting", async (tag) => {
      const payload = tag.payload;
      if (tag.type === 9 && payload.length > 0 && (payload[0] & 0x0f) === 7) {
        if (payload.length > 1 && payload[1] === 0) {
          if (!wroteVideoConfig) {
            await writeTag(outputHandle, 9, 0, payload);
            wroteVideoConfig = true;
          }
        } else if (payload.length >= 5) {
          if (videoSeen < info.videoPrefixSkip) {
            videoSeen += 1;
            return;
          }
          if (!started && (payload[0] >> 4) !== 1) {
            videoSeen += 1;
            return;
          }
          started = true;
          await writeTag(outputHandle, 9, Math.max(0, tag.timestamp - info.base), payload);
          videoSeen += 1;
        }
      } else if (tag.type === 8 && payload.length > 0 && (payload[0] >> 4) === 10) {
        if (payload.length > 1 && payload[1] === 0) {
          if (!wroteAudioConfig) {
            await writeTag(outputHandle, 8, 0, payload);
            wroteAudioConfig = true;
          }
        } else if (payload.length > 1) {
          if (audioSeen < info.audioPrefixSkip) {
            audioSeen += 1;
            return;
          }
          if (!started) {
            audioSeen += 1;
            return;
          }
          await writeTag(outputHandle, 8, Math.max(0, tag.timestamp - info.base), payload);
          audioSeen += 1;
        }
      }
    });
  } finally {
    await outputHandle.close();
  }
}

function runFfmpeg(args) {
  return new Promise((resolve, reject) => {
    const child = spawn("ffmpeg", args, { stdio: "inherit" });
    child.on("error", reject);
    child.on("exit", (code, signal) => {
      if (code === 0) {
        resolve();
      } else {
        reject(new Error(`ffmpeg exited with ${signal || `code ${code}`}`));
      }
    });
  });
}

async function probeOutput(outputPath) {
  const child = spawn("ffprobe", [
    "-v", "error", "-show_streams", "-of", "json", outputPath,
  ], { stdio: ["ignore", "pipe", "pipe"] });

  let stdout = "";
  child.stdout.setEncoding("utf8");
  child.stdout.on("data", (chunk) => {
    stdout += chunk;
  });

  try {
    const [code] = await Promise.all([
      new Promise((resolve, reject) => {
        child.on("error", reject);
        child.on("exit", resolve);
      }),
      new Promise((resolve) => {
        child.stderr.on("end", resolve);
        child.stderr.resume();
      }),
    ]);
    if (code !== 0) {
      return null;
    }

    const streams = JSON.parse(stdout).streams || [];
    const video = streams.find((stream) => stream.codec_type === "video");
    const audio = streams.find((stream) => stream.codec_type === "audio");
    if (!video || !audio) {
      return null;
    }
    return [Number.parseFloat(video.duration || 0), Number.parseFloat(audio.duration || 0)];
  } catch (error) {
    child.kill();
    return null;
  }
}

async function repair(source, output) {
  const info = await inspectTags(source);
  console.log(
    `valid audio packets: ${info.audioPackets}, valid video packets: ${info.videoPackets}, sample rate: ${info.sampleRate}`,
  );

  const tempDir = await fsp.mkdtemp(path.join(path.dirname(output), "fix_flv_stream-"));
  const cleanFlv = path.join(tempDir, `${path.basename(source, path.extname(source))}.clean.flv`);
  try {
    await writeCleanFlv(source, cleanFlv, info);
    await runFfmpeg([
      "-hide_banner",
      "-v", "warning",
      "-stats",
      "-i", cleanFlv,
      "-map", "0:v:0",
      "-map", "0:a:0",
      "-c", "copy",
      "-movflags", "+faststart",
      "-y", output,
    ]);

    const durations = await probeOutput(output);
    if (!durations) {
      return false;
    }
    const [videoDuration, audioDuration] = durations;
    console.log(`video duration: ${videoDuration.toFixed(3)}s, audio duration: ${audioDuration.toFixed(3)}s`);
    return Math.abs(videoDuration - audioDuration) < 0.5;
  } finally {
    await fsp.rm(tempDir, { recursive: true, force: true });
  }
}

async function iterFlvFiles(directory) {
  const files = [];
  async function walk(currentPath) {
    const entries = await fsp.readdir(currentPath, { withFileTypes: true });
    for (const entry of entries) {
      const entryPath = path.join(currentPath, entry.name);
      if (entry.isDirectory()) {
        await walk(entryPath);
      } else if (entry.isFile() && path.extname(entry.name).toLowerCase() === ".flv") {
        files.push(entryPath);
      }
    }
  }

  await walk(directory);
  return files.sort();
}

async function removeOutput(output) {
  try {
    await fsp.unlink(output);
  } catch (error) {
    if (error.code !== "ENOENT") {
      throw error;
    }
  }
}

async function repairOne(source, output, overwrite) {
  try {
    await fsp.access(output);
    if (!overwrite) {
      console.log(`skip: ${path.basename(output)} already exists`);
      return true;
    }
  } catch (error) {
    if (error.code !== "ENOENT") {
      throw error;
    }
  }

  try {
    if (!(await repair(source, output))) {
      await removeOutput(output);
      return false;
    }
  } catch (error) {
    await removeOutput(output);
    console.error(`error: ${error.message}`);
    return false;
  }
  return true;
}

async function repairDirectory(source, outputRoot, overwrite) {
  await fsp.mkdir(outputRoot, { recursive: true });
  const files = await iterFlvFiles(source);
  if (files.length === 0) {
    console.log("No FLV files found.");
    return 0;
  }

  const failed = [];
  for (const sourceFile of files) {
    const parsed = path.parse(sourceFile);
    const outputFile = path.join(outputRoot, `${parsed.name}_streamcopy.mp4`);
    console.log(`\n${sourceFile}`);
    if (!(await repairOne(sourceFile, outputFile, overwrite))) {
      failed.push(sourceFile);
    }
  }

  if (failed.length > 0) {
    console.log(`\n${failed.length} file(s) failed:`);
    for (const sourceFile of failed) {
      console.log(`  ${sourceFile}`);
    }
    return 1;
  }
  return 0;
}

function usage() {
  process.stderr.write([
    "usage: fix_flv_stream.js [--overwrite] source [output]",
    "",
    "Stream-parse an FLV, drop invalid tags, and stream-copy A/V into MP4.",
  ].join("\n") + "\n");
}

async function main() {
  const args = process.argv.slice(2);
  let overwrite = false;
  const positional = [];
  for (let index = 0; index < args.length; index += 1) {
    const argument = args[index];
    if (argument === "--overwrite") {
      overwrite = true;
    } else if (argument === "--help" || argument === "-h") {
      usage();
      return 0;
    } else if (argument.startsWith("--")) {
      throw new Error(`unrecognized argument: ${argument}`);
    } else {
      positional.push(argument);
    }
  }

  if (positional.length < 1 || positional.length > 2) {
    usage();
    process.stderr.write("error: expected one source and optional output\n");
    return 2;
  }

  const source = path.resolve(positional[0].replace(/^~(?=$|\/)/, os.homedir()));
  const sourceStat = await fsp.stat(source);

  if (sourceStat.isFile()) {
    const output = positional[1]
      ? path.resolve(positional[1].replace(/^~(?=$|\/)/, os.homedir()))
      : path.join(path.dirname(source), `${path.basename(source, path.extname(source))}_streamcopy.mp4`);
    return (await repairOne(source, output, overwrite)) ? 0 : 1;
  }

  if (sourceStat.isDirectory()) {
    const outputRoot = positional[1]
      ? path.resolve(positional[1].replace(/^~(?=$|\/)/, os.homedir()))
      : source;
    return await repairDirectory(source, outputRoot, overwrite);
  }

  usage();
  process.stderr.write(`error: not a file or directory: ${source}\n`);
  return 2;
}

module.exports = { iterTags, inspectTags };

if (require.main === module) {
  main().then((code) => {
    process.exitCode = code;
  }).catch((error) => {
    console.error(`error: ${error.message}`);
    process.exitCode = 2;
  });
}
